File: app/src/main/java/com/matchthree/game/engine/MatchDetector.kt
package com.matchthree.game.engine

import com.matchthree.game.model.Board
import com.matchthree.game.model.Gem
import com.matchthree.game.model.Position
import com.matchthree.game.model.Special

/** Scans a board for maximal runs of 3+ same-type gems (horizontal and vertical). */
object MatchDetector {

    fun findMatches(board: Board): List<Match> {
        val matches = mutableListOf<Match>()
        findRuns(board, horizontal = true, matches)
        findRuns(board, horizontal = false, matches)
        return matches
    }

    private fun findRuns(board: Board, horizontal: Boolean, out: MutableList<Match>) {
        val outer = if (horizontal) board.height else board.width
        val inner = if (horizontal) board.width else board.height
        for (line in 0 until outer) {
            var start = 0
            while (start < inner) {
                val first = gemAt(board, line, start, horizontal) ?: run {
                    start++
                    continue
                }
                // M4: Hypercubes are colorless and can never be part of a run.
                if (containsHypercube(first)) {
                    start++
                    continue
                }
                var end = start + 1
                while (end < inner && !containsHypercube(gemAt(board, line, end, horizontal)) &&
                    gemAt(board, line, end, horizontal)?.type == first.type
                ) {
                    end++
                }
                if (end - start >= 3) {
                    out += Match((start until end).map { index ->
                        if (horizontal) Position(line, index) else Position(index, line)
                    })
                }
                start = end
            }
        }
    }

    private fun containsHypercube(gem: Gem?): Boolean =
        gem?.special == Special.HYPERCUBE


    private fun gemAt(board: Board, line: Int, index: Int, horizontal: Boolean) =
        if (horizontal) board.gemAt(line, index) else board.gemAt(index, line)
}

File: Engine/Rules/MatchDetector.cs
using MatchThree.Engine.Model;

namespace MatchThree.Engine.Rules;

/// <summary>Scans a board for maximal runs of 3+ same-type gems (horizontal and vertical).</summary>
public static class MatchDetector
{
    /// <summary>All maximal horizontal and vertical runs on <paramref name="board"/>.</summary>
    public static List<Match> FindMatches(Board board)
    {
        var matches = new List<Match>();
        FindRuns(board, horizontal: true, matches);
        FindRuns(board, horizontal: false, matches);
        return matches;
    }

    private static void FindRuns(Board board, bool horizontal, List<Match> outList)
    {
        var outer = horizontal ? board.Height : board.Width;
        var inner = horizontal ? board.Width : board.Height;
        for (var line = 0; line < outer; line++)
        {
            var start = 0;
            while (start < inner)
            {
                var first = GemAt(board, line, start, horizontal);
                if (first is null)
                {
                    start++;
                    continue;
                }

                // Hypercubes are colorless and can never be part of a run.
                if (ContainsHypercube(first))
                {
                    start++;
                    continue;
                }

                var end = start + 1;
                while (end < inner &&
                       !ContainsHypercube(GemAt(board, line, end, horizontal)) &&
                       GemAt(board, line, end, horizontal)?.Type == first.Value.Type)
                {
                    end++;
                }

                if (end - start >= 3)
                {
                    var positions = new List<Position>(end - start);
                    for (var index = start; index < end; index++)
                    {
                        positions.Add(horizontal ? new Position(line, index) : new Position(index, line));
                    }
                    outList.Add(new Match(positions));
                }

                start = end;
            }
        }
    }

    private static bool ContainsHypercube(Gem? gem) => gem?.Special == Special.Hypercube;

    private static Gem? GemAt(Board board, int line, int index, bool horizontal) =>
        horizontal ? board.GemAt(line, index) : board.GemAt(index, line);
}